### PR TITLE
Distributed XGBoost

### PR DESCRIPTION
**Pull request by mdymczyk**
Code might be a bit more tidied up but if we want to have it in master asap then it's already working. Needs someone besides me to try it out, though.

To test:

1) 
```
git clone --recursive git@github.com:h2oai/xgboost.git
cd xgboost
```

For Unix simply `./compile-gpu-jni.sh`
For MacOS (cluster mode doesn't work on MacOS):
```
cp make/minimum.mk ./config.mk; make -j4
cd jvm-packages
mvn -DskipTests install
```

2) put the XGBoost4J jar (from `jvm-packages/xgboost4j/target/xgboost4j-0.7-jar-with-dependencies.jar`) in h2o-3's lib directory

3) build h2o and try it out

**Caveats/TODO:**

~~1) distributed models are not communicating which each other, Rabit seems to initialize/shutdown correctly but the underlying `XGBoost.train()` isn't using it for some reason - needs a fix~~

~~2) auto rebalancing for training is temporarily disabled - the distributed scoring needs to be improved to handle non consecutive chunks on a single node (i.e. 1,3 on node1 and 2,4 on node2) otherwise resulting predictions frame will have a different order to the initial training frame~~

3) can't run multiple GPU nodes on a single machine <- not sure if needs fixing (or if it's even fixable)

~~4) single node GPU modelling with `tree_method` set to `auto, exact, approx` is failing, only `hist` works. In distributed mode all of the above work <- this seems like a gpu plugin shortcoming as XGBoost Spark has the same issue, need to talk to Rory about it.~~

5) the GPU updater is not yet distributed so doesn't work - please use the CPU backend for now

### CODE REVIEW

File: h2o-algos/src/main/java/hex/tree/xgboost/XGBoostScoreTask.java [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+                        double val = preds[i][j - 1];
+                        ncs[j].addNum(val);
+                        row[j] = val;
+                    }
+                    ncs[0].addNum(hex.genmodel.GenModel.getPrediction(row, _output._priorClassDist, null, Model.defaultThreshold(_output)));
+                    ncs[ncs.length - 1].addNum(labels[i]);
+                }
+            }
+        } catch (XGBoostError xgBoostError) {
+            xgBoostError.printStackTrace();
```
Comment (mmalohlava) [inline h2o-algos/src/main/java/hex/tree/xgboost/XGBoostScoreTask.java]: Swallowing exception?

File: h2o-algos/src/main/java/hex/tree/xgboost/XGBoostUpdateTask.java [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+                _output._sparse);
+
+        if(null == trainMat) {
+            return;
+        }
+
+        // For feature importances - write out column info
+        OutputStream os;
+        try {
+          os = new FileOutputStream("/tmp/featureMap.txt");
```
Comment (mmalohlava) [inline h2o-algos/src/main/java/hex/tree/xgboost/XGBoostUpdateTask.java]: This needs to be parameterized by at least by job name else we will end with competing overwrites of the file.

File: h2o-algos/src/main/java/hex/tree/xgboost/XGBoostGPULock.java [hunk 3]
```diff
@@ -0,0 +1,21 @@
+package hex.tree.xgboost;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class XGBoostGPULock {
```
Comment (mmalohlava) [inline h2o-algos/src/main/java/hex/tree/xgboost/XGBoostGPULock.java]: Do you need to publish this one outside of xgboost package, if i understand well this is just technical aspect of implementation. Can be package private.

Comment (mdymczyk) [inline h2o-algos/src/main/java/hex/tree/xgboost/XGBoostGPULock.java]: Good points, will fix!

Other thing is the conversion methods from frame/chunks -> DMatrix are pretty similar so I want to somehow DRY them but there are enough small differences that it's a bit annoying to do I think.

**Pull request closed without merging**